pkg/ottl/ottlfuncs/func_useragent.go

```diff
@@ -0,0 +1,46 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package ottlfuncs // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/ottlfuncs"
+import (
+	"context"
+	"fmt"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl"
+	"github.com/ua-parser/uap-go/uaparser"
+	semconv "go.opentelemetry.io/collector/semconv/v1.25.0"
+)
+
+type UserAgentArguments[K any] struct {
+	UserAgent ottl.StringGetter[K]
+}
+
+func NewUserAgentFactory[K any]() ottl.Factory[K] {
+	return ottl.NewFactory("UserAgent", &UserAgentArguments[K]{}, createUserAgentFunction[K])
+}
+
+func createUserAgentFunction[K any](_ ottl.FunctionContext, oArgs ottl.Arguments) (ottl.ExprFunc[K], error) {
+	args, ok := oArgs.(*UserAgentArguments[K])
+	if !ok {
+		return nil, fmt.Errorf("URLFactory args must be of type *URLArguments[K]")
+	}
+
+	return userAgent[K](args.UserAgent), nil
+}
+
+func userAgent[K any](userAgentSource ottl.StringGetter[K]) ottl.ExprFunc[K] { //revive:disable-line:var-naming
+	return func(ctx context.Context, tCtx K) (any, error) {
+		userAgentString, err := userAgentSource.Get(ctx, tCtx)
+		if err != nil {
+			return nil, err
+		}
+		// TODO: maybe we want to manage the regex definitions separately?
+		parser := uaparser.NewFromSaved()
```

Comment:
Do we need a new parser for each pass through the function?  Could the creation of the parser be moved to above the return function so it can be reused?

Reply:
When writing the function I wasn't sure what would be the lifetime of the parser compared to the evaluation context or the collector lifetimes.

As a first implementation I went for the safe route of initializing a new parser at each invocation (with the obvious performance penalty); since we are not allowing the user to pass UA string definitions/regexes at the moment we can move the parser to a package var with a `sync.Once.Do()` to initialize it lazily at the first UA parsing and reuse that parser for any following invocation of `UserAgent(...)` function

Reply:
Since we not providing our own regexes yet, I think the second option is best.  Inside `userAgent`, before the returned `ottl.ExprFunc`, we can initialize the parser.  This is the strategy we use for regex.

Reply:
If I understood correctly option 2 should look like this ---> 095d8e44ec90f488b1436d63a9cd6d5120e3a089